src/core/memory.cc

```diff
@@ -52,6 +52,45 @@
 // Spans have a size that is a power-of-two and are naturally aligned (aka buddy
 // allocator)
 
+#ifdef SEASTAR_MODULE
+module;
+#endif
+
+#include <cassert>
+#include <unordered_set>
+#include <iostream>
+#include <optional>
+#include <thread>
+#include <memory_resource>
+
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+
+#include <boost/container/static_vector.hpp>
+
+#include <dlfcn.h>
+
+#ifndef SEASTAR_DEFAULT_ALLOCATOR
+#include <new>
+#include <cstdint>
+#include <algorithm>
+#include <limits>
+#include <cassert>
+#include <atomic>
+#include <mutex>
+#include <functional>
+#include <cstring>
+#include <boost/intrusive/list.hpp>
+#include <sys/mman.h>
+
+#ifdef SEASTAR_HAVE_NUMA
+#include <numaif.h>
+#endif
+#endif // !defined(SEASTAR_DEFAULT_ALLOCATOR)
+
+#ifdef SEASTAR_MODULE
+module seastar;
+#else
```

Comment:
Let's have something like SEASTAR_CONCEPT() that can wrap a statement.

Reply:
IIUC, what you are proposing is something like
```c++
/// util/modules.hh
#ifdef SEASTAR_CXX_MODULE
#  define SEASTAR_MODULE(x...) x
#else
#  define SEASTAR_MODULE(x...)
#endif

/// core/memory.cc
#include <seastar/util/modules.hh>
SEASTAR_MODULE(module);

/// third-party headers

SEASTAR_MODULE(module seastar);

#ifndef SEASTAR_CXX_MODULE
/// seastar  headers here
#endif
```

the upside is that we can drop some of `#if ... #endif` guards from the .cc files, but the downside is that we are hiding more C++ keywords, and we need to have the `seastar/util/modules.hh` header sitting at the top of every .cc file. IMHO, it looks more weird than the `#ifdef ... #endif` pairs, but if you insist, i will update all .cc file to reflect this change.

Reply:
We wouldn't be hiding any keywords, since the macros don't supply keywords themselves. They only replace #ifdef/#endif pairs.

I don't insist though. For concepts it really helps because the concept is part of a definition, here it's just decoration at the top/bottom of a file.